ext/src/http/client/curl/http_client_curl.cc

```diff
@@ -48,10 +53,53 @@ void Session::SendRequest(
     reuse_connection = session_id_ % http_client_.GetMaxSessionsPerConnection() != 0;
   }
 
+  if (http_request_->compression_ == opentelemetry::ext::http::client::Compression::kGzip)
+  {
+#ifdef ENABLE_OTLP_COMPRESSION_PREVIEW
+    http_request_->AddHeader("Content-Encoding", "gzip");
+
+    opentelemetry::ext::http::client::Body compressed_body(http_request_->body_.size());
+    z_stream zs;
+    zs.zalloc    = Z_NULL;
+    zs.zfree     = Z_NULL;
+    zs.opaque    = Z_NULL;
+    zs.avail_in  = static_cast<uInt>(http_request_->body_.size());
+    zs.next_in   = http_request_->body_.data();
+    zs.avail_out = static_cast<uInt>(compressed_body.size());
+    zs.next_out  = compressed_body.data();
+
+    // ZLIB: Have to maually specify 16 bits for the Gzip headers
+    const int window_bits = 15 + 16;
+
+    int stream =
+        deflateInit2(&zs, Z_DEFAULT_COMPRESSION, Z_DEFLATED, window_bits, 8, Z_DEFAULT_STRATEGY);
```

Comment:
@perhapsmaple  - Do you think we can use the in-place compression of the data without using a separate output buffer `compressed_body`, as suggested here - https://stackoverflow.com/questions/12398377/is-it-possible-to-have-zlib-read-from-and-write-to-the-same-memory-buffer/12412863#12412863. This is suggested by "Mark Adler", the owner of zlib. 
If not feasible in this PR., good to add a TODO and create an issue for tracking, so some one can pick it up.